utils_cv/action_recognition/dataset.py

```diff
@@ -0,0 +1,473 @@
+        assert sample_step > 0
+        assert num_segments > 0
+
+        if temporal_jitter:
+            assert temporal_jitter_step > 0
+
+        if train_split_file:
+            assert Path(train_split_file).exists()
+            assert (
+                test_split_file is not None and Path(test_split_file).exists()
+            )
+
+        if test_split_file:
+            assert Path(test_split_file).exists()
+            assert (
+                train_split_file is not None
+                and Path(train_split_file).exists()
+            )
+
+        self.root = root
+        self.num_segments = num_segments
+        self.sample_length = sample_length
+        self.sample_step = sample_step
+        self.presample_length = sample_length * sample_step
+        self.temporal_jitter_step = temporal_jitter_step
+        self.train_transforms = train_transforms
+        self.test_transforms = test_transforms
+        self.random_shift = random_shift
+        self.temporal_jitter = temporal_jitter
+        self.batch_size = batch_size
+        self.video_ext = video_ext
+        self.warning = warning
+
+        # create training and validation datasets
+        self.train_ds, self.test_ds = (
+            self.split_with_file(
+                train_split_file=train_split_file,
+                test_split_file=test_split_file,
+            )
+            if train_split_file
+            else self.split_train_test(train_pct=train_pct)
+        )
+
+        # initialize dataloaders
+        self.init_data_loaders()
+
+    def split_train_test(
+        self, train_pct: float = 0.8
+    ) -> Tuple[Dataset, Dataset]:
+        """ Split this dataset into a training and testing set
+
+        Args:
+            train_pct: the ratio of images to use for training vs
+            testing
+
+        Return
+            A training and testing dataset in that order
+        """
+        pass
+
+    def split_with_file(
+        self,
+        train_split_file: Union[Path, str],
+        test_split_file: Union[Path, str],
+    ) -> Tuple[Dataset, Dataset]:
+        """ Split this dataset into a training and testing set using a split file.
+
+        Args:
+            split_files: a tuple of 2 files
+
+        Return:
+            A training and testing dataset in that order
+        """
+        self.video_records = []
+
+        # add train records
+        self.video_records.extend(
+            [VideoRecord(x.strip().split(" ")) for x in open(train_split_file)]
+        )
+        train_len = len(self.video_records)
+
+        # add validation records
+        self.video_records.extend(
+            [VideoRecord(x.strip().split(" ")) for x in open(test_split_file)]
+        )
+
+        # create indices
+        indices = torch.arange(0, len(self.video_records))
+        train_range = indices[:train_len]
+        test_range = indices[train_len:]
+
+        # create train subset
+        train = copy.deepcopy(Subset(self, train_range))
```

Comment:
[question] Remind me, we do these deepcopies to create a structure similar to fast.ai?

Reply:
hmm dont remember how it related to fastai. but for object detection, we used the same pattern so that the dataset object could hold separate copies for the train and validation sets, each with their own parameters.